src/chromo/common.py

```diff
@@ -543,6 +537,45 @@ def __getnewargs__(self):
         # upon unpickling, create EventData object instead of MCEvent object
         return (EventData,)
 
+    def _add_init_beam_info(self):
+        pass
+
+    def _fill_initial_beam(self):
+        beam = self.kin._get_beam_data(self._generator_frame)
+        for field, beam_field in beam.items():
+            event_field = getattr(self, field)
+            if event_field is not None:
+                if field == "mothers":
+                    # all mothers that looks like [-1, -1], [0, -1], [1, -1], [1, 1]
+                    # are changed to [0, 1], i.e. attach to a common vertex
+                    # of to beam particles
+                    event_field[
+                        np.isin(event_field, [[-1, -1], [0, -1], [1, -1], [1, 1]]).all(
+                            axis=1
+                        )
+                    ] = [0, 1]
+
+                event_field[0:2] = beam_field
+
+    def _append_initial_beam(self):
+        beam = self.kin._get_beam_data(self._generator_frame)
+        for field, beam_field in beam.items():
+            event_field = getattr(self, field)
+            if field in ["mothers", "daughters"]:
+                if event_field is None:
+                    continue
```

Comment:
Shouldn't that be before the if?

Reply:
Yes, it is also possible. I moved it up.